### PR TITLE
Player leaves the game if he changes the world.

### PR DESCRIPTION
**Pull request by Ybrin**
This function is needed if a player types in a command like /lobby and
is currently playing Bedwars. He will be automatically removed from that
game.

For our server, we want to have a basic command (/lobby) for all minigames. Currently we wrote a simple plugin which listens to the PlayerChangedWorldEvent and triggers the command /bw leave.
The problem is, this method is not always working because it sometimes triggers too early or too late.
I think this option could be also useful for others, so I am sharing my solution.

Basically if you change your world with a command other than /bw leave, you'll automatically leave the game you are currently in.

### CODE REVIEW

**Comment by Ybrin** [comment]
You can change the worl from whereever you want. This code is just responsible if you are in a Bedwars game and you leave that world. This code won't do anything except that.

**Comment by Ybrin** [comment]
I found an error. I try to fix it, don't accept this pull request until I am done...

**Comment by Yannici** [comment]
Ok, I'm waiting for your final commit.

**Comment by Ybrin** [comment]
Ok I fixed the bug. It should now work as expected.

This is a simple test: https://vid.me/DY0t

**Comment by Yannici** [comment]
It's nice it is working, but I can't merge it. I don't want such ugly and unclean code in my project :(

**Comment by Ybrin** [comment]
I could also save a Boolean as metadata to the Player and create a new method which automates the process but normally I hat working with Metadata as they are just available for Minecraft projects and I don't want to rely on such things.

If you would accept it later, I would change it now.

**Comment by Yannici** [comment]
Basically the problem I have with you code is, that there is no need to store any data in any list/hashmap nor metadata of the player! Every information you need you have in the `Game` object.

**Comment by Ybrin** [comment]
Your `Game` class is very confusing. Without javadocs it is hard to understand such large blocks of code. I go though your code again and try to use your existing methods. 
Could you just say me what the method is called which returns a value whether the player is currently leaving or joining?

Until this was merged I'll use my version because this is really something that is needed for our server (and also for other servers).

**Comment by Ybrin** [comment]
You indeed need a value which stores whether the player is leaving the game or joining. If you don't do it like that, the plugin will always try to remove a player twice from a game if he leaves without the `/lobby` command.
I am sorry but I don't know how you would like your code to look like (there is no contributing file in your project).
Also, I haven't found a method which returns whether the Player is leaving or joining. I could create one but it is only needed in the PlayerListener class so why would you do something inside your Game class? The latter is long enough and should not be unnecessarily expanded.

**Comment by Yannici** [comment]
No you don't need to store anything. 

If the game is running and a player is switching the world, the player can be kicked. When game ends the `isEndGameRunning()` of the `GameCycle` will be set to true, so you know when the player will be switch world on game ending. When player is joining the GameState is still `WAITING`, so you know he is joining ;-) So easy ...

I don't like static fields, that's it. And I don't need a guidline for other programmers. If they study my code, they know how I like my code. I prefer pure OOP and static fields only in case of emergency ...

**Comment by Ybrin** [comment]
Thank you for your answer. I know the preferences are different and every programmer has his own way to "style" his code.

I am sorry but I tried to fix this as soon as possible and for that reason I didn't spend much time studying your code so I have done it my way.

I change this to OOP and use your methods (which I haven't seen yet) and then I will contact you again.

**Comment by Ybrin** [comment]
I have now done exactly what you said. The problem is, like I wanted to explain, now the player can't type `/lobby` if he is still in the lobby (because the GameState is still WAITING). If I change the code, we would have the same problem as explained before, if the game starts.

The only possibility I see other that my solution (or storing a Metadata value) is to make a new GameState like `TELEPORTING` which will be set while all the players are teleported from the lobby to the game.

I don't know if you want me to change something so important in your code but the newest commit works well if you don't want to use such commands in the lobby but just if a game is running.

**Comment by Ybrin** [comment]
This is not the problem, I CAN be teleported to the lobby with the command `/lobby` but I will still be in the Bedwars game because I have never left it. I am trying to add exactly this feature.

**Comment by Ybrin** [comment]
Just read my first post, I have done this but you can't time it to be as exact as a native implementation. And also, other people have asked this before on the Spigot wiki. I tried to help and the solution is indeed working but for now just while the game is not WAITING.

**Comment by Yannici** [comment]
Thanks @Ybrin for your involvement!
You're right, it's definitely difficult to know when the player is switching world from lobby caused by a command or caused by the game start. The only possible solutions (but not 100% perfect) I found are the following:

1. Check whether the player is switching the world to game world and check if the player comes from the lobby world. As long as the lobby (the `/lobby` lobby) is not in the game world, this works fine!
2. On world switching, check if the game is currently starting (maybe by checking the Countdown state or something...)

Otherwise you're right, you have to store the information that the player will leave in the lobby by a command (`/lobby`) and kick him out of the game. I prefer storing that information in the game-class with a `Set` or `List`. Or a metadata of the player (which is basically the easiest way).

**Comment by Ybrin** [comment]
Storing the information is more accurate so I'll do it that way. You'll definitely hear from me soon.

**Comment by Ybrin** [comment]
I used the PlayerSettings class to store a `isTeleporting` boolean value. If a player is teleported somewhere other than his current world, this value will be set to `true`, if he uses a command like `/lobby` this value is not true and he will be kicked.

I hope you can now merge my code because, like I said before, this is a feature many people want to have in every Minigame.

**Comment by Ybrin** [comment]
@Block-Build I think you should read all the comments again...

**Comment by Ybrin** [comment]
Kein Kommentar...

**Comment by Yannici** [comment]
@Block-Build Vergiss es einfach ;-) Ich glaube du weißt nicht worum es hier geht.
@Ybrin Thanks! Now it's like I want it to have. I merged it right now. Thanks for your improvement.

**Comment by Yannici** [comment]
I made a little code optimization in https://github.com/Yannici/bedwars-reloaded/commit/9ed311a58472f72b06c546abd22288cff17b9a2c

**Comment by Ybrin** [comment]
@Yannici Kein Problem, ach ja ich kann außerdem auch Deutsch aber ich bin der Meinung dass Komentare auf GitHub Englisch sein sollten deswegen habe ich so begonnen...

**Comment by Yannici** [comment]
True :D
